### PR TITLE
Add support for OpenAI API compatible models

### PR DESCRIPTION
**Pull request by chenhunghan**
Many popular open-sourced project offers OpenAI API compatible endpoints, for example
- [llama.cpp](https://github.com/ggerganov/llama.cpp/tree/master/examples/server) (41K ⭐)
- [oobabooga/text-generation-webui](https://github.com/oobabooga/text-generation-webui/tree/main/extensions/openai) (23.9K ⭐)
- [LocalAI](https://github.com/go-skynet/LocalAI) (11.2K ⭐)
- [FastChat](https://github.com/lm-sys/FastChat/blob/main/docs/openai_api.md) (27.5K ⭐)
- [llama-cpp-python](https://github.com/abetlen/llama-cpp-python) (3.3K ⭐)
- [ialacol](https://github.com/chenhunghan/ialacol) (82 ⭐)

This PR adds OpenAI API endpoint compatibility, if merge, `chat-ui` can be used as the chat user interface for any project mentioned above. The official OpenAI endpoints are also support, user can quickly evaluate GPT3/GPT4 and/or any OSS models (GGUF/GPTQ...) hosted by for example [llama.cpp](https://github.com/ggerganov/llama.cpp/tree/master/examples/server).

The difference between this PR and https://github.com/huggingface/chat-ui/pull/443 is
- Model base selection not endpoint base, can have multiple models at the same time and select one via the dropdown.
- Robust SSE event decoding via [OpenAI node](https://github.com/openai/openai-node/blob/master/src/streaming.ts#L77)
- Better support for system prompt, and all the OpenAI supported parameters.

### CODE REVIEW

**Comment by julien-blanchon** [comment]
Using:

```json
"userMessageToken": "### User:\n",
"userMessageEndToken": "\n",
"assistantMessageToken": "### Assistant:\n",
```

Could lead to easy prompt injection

Screenshot 2023-09-20 at 15 02 17

**Comment by julien-blanchon** [comment]
Expect that it seems that everything work fine

File: README.md
Comment (julien-blanchon) [reply README.md:176]: Yes prompt injection are quite easy to do here

File: README.md
Comment (julien-blanchon) [inline README.md:188]: This lead to the message to stop when hitting a subsubtitle in markdown, this is quite common

**Review by julien-blanchon — COMMENTED** [review]
Everything work fine ! However they are some "too" easy prompt injection

File: README.md
Comment (chenhunghan) [reply README.md:176]: Good point, this is just a random example that was used by a llama2 model I was hacking, probably a bad example. I will add a new one.

File: .env
Comment (julien-blanchon) [reply .env:11]: Yes make sense !

File: src/lib/server/models.ts
Comment (chenhunghan) [reply src/lib/server/models.ts]: Actually `OpenAI` instance would throw errors (by design) if there is no key provided, so we still need to provide `sk-` in some way as default to prevent this behaviour. https://github.com/huggingface/chat-ui/pull/452/commits/d9509668bbe4550ee331d31f332c814b748c68f1

**Review by julien-blanchon — APPROVED** [review]

**Comment by gururise** [comment]
Been trying this and with this config:
```
  {
    "name": "GPT-3.5 Turbo",
    "id": "gpt-3.5-turbo",
    "userMessageToken": "User:\n",
    "userMessageEndToken": "\n",
    "assistantMessageToken": "Assistant:\n",
    "preprompt": "You are a helpful assistant",
    "promptExamples": [],
    "parameters": {
      "temperature": 0.9,
      "top_p": 0.95,
      "repetition_penalty": 1.2,
      "top_k": 50,
      "truncate": 1000,
      "max_new_tokens": 1024,
      "stop": []
    },
    "endpoints": [{
      "host" : "openai-compatible"
    }]
  }
```

inference seems to get stuck.  Once I shut down the node server and restart it, it will run inference on the last text issued by the user:

image

**Comment by chenhunghan** [comment]
> "name": "GPT-3.5 Turbo"


> Been trying this and with this config:
> 
> ```
>   {
>     "name": "GPT-3.5 Turbo",
>     "id": "gpt-3.5-turbo",
>     "userMessageToken": "User:\n",
>     "userMessageEndToken": "\n",
>     "assistantMessageToken": "Assistant:\n",
>     "preprompt": "You are a helpful assistant",
>     "promptExamples": [],
>     "parameters": {
>       "temperature": 0.9,
>       "top_p": 0.95,
>       "repetition_penalty": 1.2,
>       "top_k": 50,
>       "truncate": 1000,
>       "max_new_tokens": 1024,
>       "stop": []
>     },
>     "endpoints": [{
>       "host" : "openai-compatible"
>     }]
>   }
> ```
> 
> inference seems to get stuck. Once I shut down the node server and restart it, it will run inference on the last text issued by the user:
> 
> image
> 
> If I wait long enough, I get this error:
> 
> ```
> Error: {"error":"Internal Server Error"}
>     at Module.generateFromDefaultEndpoint (/home/gene/Downloads/chat-ui/src/lib/server/generateFromDefaultEndpoint.ts:129:11)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async Module.summarize (/home/gene/Downloads/chat-ui/src/lib/server/summarize.ts:17:26)
>     at async saveLast (/home/gene/Downloads/chat-ui/src/routes/conversation/[id]/+server.ts:174:24)
> ```

I tried the same setting, but can't reproduce, there seems exceptions when calling `generateFromDefaultEndpoint`, this was for generating a summary of the coversation, anyway, I added a [try...catch there](https://github.com/huggingface/chat-ui/pull/452/commits/fac7113c6a88558f6e8a77a1f3532952fa9ae54e), would you mind try again? You will be able to see the real errors.

**Comment by chenhunghan** [comment]
@nsarrazin @philschmid PTAL.

**Comment by gururise** [comment]
> > "name": "GPT-3.5 Turbo"
> > ...
> > If I wait long enough, I get this error:
> > ```
> > Error: {"error":"Internal Server Error"}
> >     at Module.generateFromDefaultEndpoint (/home/gene/Downloads/chat-ui/src/lib/server/generateFromDefaultEndpoint.ts:129:11)
> >     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> >     at async Module.summarize (/home/gene/Downloads/chat-ui/src/lib/server/summarize.ts:17:26)
> >     at async saveLast (/home/gene/Downloads/chat-ui/src/routes/conversation/[id]/+server.ts:174:24)
> > ```
> 
> I tried the same setting, but can't reproduce, there seems exceptions when calling `generateFromDefaultEndpoint`, this was for generating a summary of the coversation, anyway, I added a [try...catch there](https://github.com/huggingface/chat-ui/pull/452/commits/fac7113c6a88558f6e8a77a1f3532952fa9ae54e), would you mind try again? You will be able to see the real errors.

Sorry I haven't been able to replicate it.

**Comment by gururise** [comment]
Trying to use the LiteLLM-Proxy and Together.ai model:
```
{
    "name": "vicuna-1.5",
    "id": "together_ai/lmsys/vicuna-13b-v1.5-16k",
    "userMessageToken": "User:\n",
    "userMessageEndToken": "\n",
    "assistantMessageToken": "Assistant:\n",
    "preprompt": "You are a helpful assistant",
    "promptExamples": [],
    "parameters": {
      "temperature": 0.9,
      "top_p": 0.95,
      "repetition_penalty": 1.2,
      "top_k": 50,
      "truncate": 1000,
      "max_new_tokens": 1024,
      "stop": []
    },
    "endpoints": [{
      "host" : "openai-compatible",
      "baseUrl": "http://localhost:8081",
      "type": "chat_completions"
    }]
  }
```

The LiteLLM endpoint (**http://localhost:8081/chat/completions**) never gets called, and I see this error message in the browser console:
```
TypeError: NetworkError when attempting to fetch resource. [+page.svelte:154:11](http://localhost:5173/src/routes/conversation/[id]/+page.svelte)
    writeMessage
```

**Comment by chenhunghan** [comment]
> ```
> TypeError: NetworkError when attempting to fetch resource. [+page.svelte:154:11](http://localhost:5173/src/routes/conversation/[id]/+page.svelte)
>     writeMessage
> ```

There was a typo in READE, should be `baseURL` instead of `baseUrl`, update your `.env.local` it should send the request to LiteLLM.

**Comment by gururise** [comment]
> > ```
> > TypeError: NetworkError when attempting to fetch resource. [+page.svelte:154:11](http://localhost:5173/src/routes/conversation/[id]/+page.svelte)
> >     writeMessage
> > ```
> 
> There was a [typo](https://github.com/huggingface/chat-ui/pull/452/commits/d6225ad6aa812357b1dbc8d8312861398293a669) in README, should be `baseURL` instead of `baseUrl`, update your `.env.local` it should send the request to LiteLLM.

Even after changing to "baseURL", I am still getting the same network error and I can see that the endpoint is not being accessed:

```
{
    "name": "vicuna-1.5",
    "id": "together_ai/lmsys/vicuna-13b-v1.5-16k",
    "userMessageToken": "User:\n",
    "userMessageEndToken": "\n",
    "assistantMessageToken": "Assistant:\n",
    "preprompt": "You are a helpful assistant",
    "promptExamples": [],
    "parameters": {
      "temperature": 0.9,
      "top_p": 0.95,
      "repetition_penalty": 1.2,
      "top_k": 50,
      "truncate": 1000,
      "max_new_tokens": 1024,
      "stop": []
    },
    "endpoints": [{
      "host" : "openai-compatible",
      "baseURL": "http://localhost:8081",
      "type": "chat_completions"
    }]
  }
```

**Comment by chenhunghan** [comment]
> Even after changing to "baseURL", I am still getting the same network error and I can see that the endpoint (http://localhost:8081/chat/completions) is not being accessed:

How did you run `chat-ui`? Did you restart `chat-ui` to let it parse the `.env.local` again? Will be helpful to know what goes wrong if you can add a `try...catch` around `await openai.chat.completions.create` in `src/routes/conversation/[id]/+server.ts` like

Screenshot 2023-09-23 at 20 34 53

**Comment by gururise** [comment]
Yes, restarted chat-ui.

This is the error on the console.
```
APIConnectionError: undefined Connection error.
    at OpenAI.makeRequest (file:///home/gene/Downloads/chat-ui/node_modules/openai/core.mjs:292:13)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async openAITextGenerationStream (/home/gene/Downloads/chat-ui/src/routes/conversation/[id]/+server.ts:209:13)
    at async Object.start (/home/gene/Downloads/chat-ui/src/routes/conversation/[id]/+server.ts:225:73) {
  status: undefined,
  headers: undefined,
  error: undefined,
  code: undefined,
  param: undefined,
  type: undefined,
  cause: FetchError: request to http://localhost:8081/chat/completions failed, reason: connect ECONNREFUSED ::1:8081
      at ClientRequest.<anonymous> (/home/gene/Downloads/chat-ui/node_modules/node-fetch/lib/index.js:1501:11)
      at ClientRequest.emit (node:events:513:28)
      at Socket.socketErrorListener (node:_http_client:502:9)
      at Socket.emit (node:events:525:35)
      at emitErrorNT (node:internal/streams/destroy:151:8)
      at emitErrorCloseNT (node:internal/streams/destroy:116:3)
      at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
    type: 'system',
    errno: 'ECONNREFUSED',
    code: 'ECONNREFUSED'
  }
}
```

**Comment by gururise** [comment]
**UPDATE:** I figured out the problem!
The baseURL cannot be **localhost**. OpenAI library does not like this and will automatically refuse connection.

Changing the baseURL to 127.0.0.1 and everything works!  Thank You!

**Comment by gururise** [comment]
Think I may have found a bug.

Here is my models in .env.local
```
MODELS=`[
  {
    "name": "Vicuna",
    "id": "together_ai/Austism/chronos-hermes-13b",
    "userMessageToken": "### Instruction: ",
    "userMessageEndToken": "\n\n",
    "assistantMessageToken": "### Response: ",
    "assistantMessageEndToken": "\n\n",
    "preprompt": "You are a helpful assistant.",
    "promptExamples": [],
    "parameters": {
      "temperature": 0.9,
      "top_p": 0.95,
      "repetition_penalty": 1.2,
      "top_k": 50,
      "truncate": 1000,
      "max_new_tokens": 1024,
      "stop": ["</s>"]
    },
    "endpoints": [{
      "host" : "openai-compatible",
      "baseURL": "http://127.0.0.1:8081/",
      "type": "chat_completions"
    }]
  },
  {
    "name": "GPT-3.5 Turbo",
    "id": "gpt-3.5-turbo",
    "userMessageToken": "User:\n",
    "userMessageEndToken": "\n",
    "assistantMessageToken": "Assistant:\n",
    "preprompt": "You are a helpful assistant",
    "promptExamples": [],
    "parameters": {
      "temperature": 0.9,
      "top_p": 0.95,
      "repetition_penalty": 1.2,
      "top_k": 50,
      "truncate": 1000,
      "max_new_tokens": 1024,
      "stop": []
    },
    "endpoints": [{
      "host" : "openai-compatible"
    }]
  }
]`
```

When I select GPT-3.5 Turbo as the model in the HF Chat UI, it still tries to access the litellm proxy I have running at http://127.0.0.1:8001

Since I'm using my open ai key, I get a Authentication Error (since I have a different key for the litellm proxy).

If I **SWAP** the positions of how I define the models so GPT-3.5 Turbo is declared first, it works fine.

**Comment by chenhunghan** [comment]
> Changing the baseURL to 127.0.0.1 and everything works! Thank You!

Likely related to how LiteLLM bind to the _localhost_, they might have only bind to `127.0.0.1`. Anyway glad to hear to solve the problem.

**Comment by chenhunghan** [comment]
> Think I may have found a bug.
> 
> Here is my models in .env.local
> ...
> When I select GPT-3.5 Turbo as the model in the HF Chat UI, it still tries to access the litellm proxy I have running at http://127.0.0.1:8001
> 
> Since I'm using my open ai key, I get a Authentication Error (since I have a different key for the litellm proxy).
> 
> If I **SWAP** the positions of how I define the models so GPT-3.5 Turbo is declared first, it works fine.

You can specify the API Key you want to use per model. For example 
```
MODELS=`[
  {
    "name": "Vicuna",
    "id": "together_ai/Austism/chronos-hermes-13b",
    "userMessageToken": "### Instruction: ",
    "userMessageEndToken": "\n\n",
    "assistantMessageToken": "### Response: ",
    "assistantMessageEndToken": "\n\n",
    "preprompt": "You are a helpful assistant.",
    "promptExamples": [],
    "parameters": {
      "temperature": 0.9,
      "top_p": 0.95,
      "repetition_penalty": 1.2,
      "top_k": 50,
      "truncate": 1000,
      "max_new_tokens": 1024,
      "stop": ["</s>"]
    },
    "endpoints": [{
      "host" : "openai-compatible",
      "baseURL": "http://127.0.0.1:8081/",
      "type": "chat_completions",
      "apiKey": # Your litellm API Key here
    }]
  },
  {
    "name": "GPT-3.5 Turbo",
    "id": "gpt-3.5-turbo",
    "userMessageToken": "User:\n",
    "userMessageEndToken": "\n",
    "assistantMessageToken": "Assistant:\n",
    "preprompt": "You are a helpful assistant",
    "promptExamples": [],
    "parameters": {
      "temperature": 0.9,
      "top_p": 0.95,
      "repetition_penalty": 1.2,
      "top_k": 50,
      "truncate": 1000,
      "max_new_tokens": 1024,
      "stop": []
    },
    "endpoints": [{
      "host" : "openai-compatible",
      "apiKey": # If you don't add a value here, it will by default use the env `OPENAI_API_KEY` 
    }]
  }
]`
```

**Comment by krrishdholakia** [comment]
is there anything blocking this PR from being merged?

**Comment by Extremys** [comment]
It seems to not pass the build phase for prod. Or maybe I'm doing something wrong...

`#25 2.704 vite v4.3.9 building SSR bundle for production...
#25 2.778 transforming...
#25 5.137 Browserslist: caniuse-lite is outdated. Please run:
#25 5.137   npx update-browserslist-db@latest
#25 5.137   Why you should do it regularly: https://github.com/browserslist/update-db#readme
#25 8.569 "BaseClient" and "TokenSet" are imported from external module "openid-client" but never used in "src/lib/server/auth.ts".
#25 8.569 ? 170 modules transformed.
#25 8.764 rendering chunks...
#25 10.35 file:///app/node_modules/zod/lib/index.mjs:537
#25 10.35                 const error = new ZodError(ctx.common.issues);
#25 10.35                               ^
#25 10.35 
#25 10.35 ZodError: [
#25 10.35   {
#25 10.35     "code": "invalid_union",
#25 10.35     "unionErrors": [
#25 10.35       {
#25 10.35         "issues": [
#25 10.35           {
#25 10.35             "received": "openai-compatible",
#25 10.35             "code": "invalid_literal",
#25 10.35             "expected": "sagemaker",
#25 10.35             "path": [
#25 10.35               0,
#25 10.35               "endpoints",
#25 10.35               0,
#25 10.35               "host"
#25 10.35             ],
#25 10.35             "message": "Invalid literal value, expected \"sagemaker\""
#25 10.35           },
#25 10.35           {
#25 10.35             "code": "invalid_type",
#25 10.35             "expected": "string",
#25 10.35             "received": "undefined",
#25 10.35             "path": [
#25 10.35               0,
#25 10.35               "endpoints",
#25 10.35               0,
#25 10.35               "url"
#25 10.35             ],
#25 10.35             "message": "Required"
#25 10.35           },
#25 10.35           {
#25 10.35             "code": "invalid_type",
#25 10.35             "expected": "string",
#25 10.35             "received": "undefined",
#25 10.35             "path": [
#25 10.35               0,
#25 10.35               "endpoints",
#25 10.35               0,
#25 10.35               "accessKey"
#25 10.35             ],
#25 10.35             "message": "Required"
#25 10.35           },
#25 10.35           {
#25 10.35             "code": "invalid_type",
#25 10.35             "expected": "string",
#25 10.35             "received": "undefined",
#25 10.35             "path": [
#25 10.35               0,
#25 10.35               "endpoints",
#25 10.35               0,
#25 10.35               "secretKey"
#25 10.35             ],
#25 10.35             "message": "Required"
#25 10.35           }
#25 10.35         ],
#25 10.35         "name": "ZodError"
#25 10.35       },
#25 10.35       {
#25 10.35         "issues": [
#25 10.35           {
#25 10.35             "code": "invalid_union",
#25 10.35             "unionErrors": [
#25 10.35               {
#25 10.35                 "issues": [
#25 10.35                   {
#25 10.35                     "received": "openai-compatible",
#25 10.35                     "code": "invalid_literal",
#25 10.35                     "expected": "tgi",
#25 10.35                     "path": [
#25 10.35                       0,
#25 10.35                       "endpoints",
#25 10.35                       0,
#25 10.35                       "host"
#25 10.35                     ],
#25 10.35                     "message": "Invalid literal value, expected \"tgi\""
#25 10.35                   }
#25 10.35                 ],
#25 10.35                 "name": "ZodError"
#25 10.35               },
#25 10.35               {
#25 10.35                 "issues": [
#25 10.35                   {
#25 10.35                     "code": "invalid_type",
#25 10.35                     "expected": "undefined",
#25 10.35                     "received": "string",
#25 10.35                     "path": [
#25 10.35                       0,
#25 10.35                       "endpoints",
#25 10.35                       0,
#25 10.35                       "host"
#25 10.35                     ],
#25 10.35                     "message": "Expected undefined, received string"
#25 10.35                   }
#25 10.35                 ],
#25 10.35                 "name": "ZodError"
#25 10.35               }
#25 10.35             ],
#25 10.35             "path": [
#25 10.35               0,
#25 10.35               "endpoints",
#25 10.35               0,
#25 10.35               "host"
#25 10.35             ],
#25 10.35             "message": "Invalid input"
#25 10.35           },
#25 10.35           {
#25 10.35             "code": "invalid_type",
#25 10.35             "expected": "string",
#25 10.35             "received": "undefined",
#25 10.35             "path": [
#25 10.35               0,
#25 10.35               "endpoints",
#25 10.35               0,
#25 10.35               "url"
#25 10.35             ],
#25 10.35             "message": "Required"
#25 10.35           }
#25 10.35         ],
#25 10.35         "name": "ZodError"
#25 10.35       }
#25 10.35     ],
#25 10.35     "path": [
#25 10.35       0,
#25 10.35       "endpoints",
#25 10.35       0
#25 10.35     ],
#25 10.35     "message": "Invalid input"
#25 10.35   }
#25 10.35 ]
#25 10.35     at get error [as error] (file:///app/node_modules/zod/lib/index.mjs:537:31)
#25 10.35     at ZodArray.parse (file:///app/node_modules/zod/lib/index.mjs:636:22)
#25 10.35     at file:///app/.svelte-kit/output/server/chunks/models.js:96:3
#25 10.35     at ModuleJob.run (node:internal/modules/esm/module_job:193:25)`

**Comment by Extremys** [comment]
Hello I am using FastChat OpenAI compatible API as backend but the Ui seems to not work properly, it's getting stuck in generation mode:

cfg:
```
  {
    "name": "vicuna-13b-v1.5",
    "id": "vicuna-13b-v1.5",
    "userMessageToken": "User:\n",
    "userMessageEndToken": "\n",
    "assistantMessageToken": "Assistant:\n",
    "preprompt": "You are a helpful assistant",
    "promptExamples": [],
    "parameters": {
      "temperature": 0.9,
      "top_p": 0.95,
      "repetition_penalty": 1.2,
      "top_k": 50,
      "truncate": 2000,
      "max_new_tokens": 2048,
      "stop": ["</s>"],
      "stream": true
    },
    "endpoints": [
  {
    "host": "openai-compatible",
    "apiKey": "{{ CHATUI_TOKEN }}",
    "baseURL": "https://myopenai-api/v1",
    "type": "chat_completions",
    "weight": 1
  }
```
image


Any idea?

**Comment by chenhunghan** [comment]
I have not received any comments from the maintainers for over a month, it seems they have no interest or there are conflicts of interests to support OpenAI-styled API. For those interests in using chat-ui with OpenAI-styled API, please follow the forked version that tried to sync with the upstream . https://github.com/ialacol/chat-ui/tree/main

**Comment by julien-blanchon** [comment]
Humm I don't think they have any conflict of interest ^^. I think they just lacking of time to review it or maybe just forgot. For me this is looking fine, it's okay to merge.

**Comment by nsarrazin** [comment]
Hey guys! No conflicts of interest, as far as I know we're pretty happy to make chat-ui backend agnostic. I just wanted to take the time to refactor things a bit so that we could add more backends easily in the future, in a way that scales. I wanted to use dynamic imports so that not everyone needs to install the `openai` or `aws4fetch` packages if they're not going to use those backends. 

I just didn't have the time to design a common backend API I was happy with, but seems like it is a pressing issue for some so I'll have a look soon 😁

**Comment by nsarrazin** [comment]
Hi everyone! Quick update on the support for openAI type endpoints. I finished my refactoring on the pr #541, I tested it using the OpenAI API and it worked well. 

I'll be testing it more thoroughly and updating the docs but feel free to do it as well with your locally hosted APIs to let me know if it works 😄 

Thanks @chenhunghan for the great work on this!

**Pull request closed without merging**
